compiler/rustc_expand/src/mbe/macro_rules.rs

```diff
@@ -651,6 +660,30 @@ fn check_matcher(sess: &ParseSess, def: &ast::Item, matcher: &[mbe::TokenTree])
     err == sess.span_diagnostic.err_count()
 }
 
+fn has_compile_error_macro(rhs: &mbe::TokenTree) -> bool {
+    use mbe::TokenTree::*;
+    match rhs {
+        Delimited(_sp, d) => {
+            let has_compile_error = d.tts.array_windows::<3>().any(|[ident, bang, args]| {
+                if let Token(ident) = ident &&
```

Comment:
I'm not sure a simple scan of the token tree is sufficient here.  Suppose a macro rule produces conditional output:

```rust
macro_rules! foo {
    (one) => {
        #[cfg(sometimes)]
        compile_error!();

        #[cfg(not(sometimes))]
        // no compile_error!
        todo!();
    };
}
```

Then shouldn't we still be linting the macro as unused when `not(sometimes)`?

Reply:
Unfortunately it's hard to do "static" analysis for macros. Macros act in a heavily untyped domain of token trees, and there are many more ways the `compile_error` macro could be invoked. See also: https://github.com/rust-lang/rust/pull/96150#issuecomment-1132816837

Reply:
It's a lint, it can use heuristics.
If detecting the sequence "`compile_error` `!`" removes more false positives than introduces false negatives, then it's fine.